source/bank_20.asm

```diff
@@ -0,0 +1,51 @@
+INCLUDE "level_constants.inc"
+INCLUDE "puzzle_constants.inc"
+
+SECTION "calc_safari_map_puzzle_index", ROMX[$46B5], BANK[$20]
+calc_safari_map_puzzle_index::
+    ld a, [w_safari_map_level]
+    cp LEVEL_JUNGLE_ZONE
+    jp z, calc_safari_map_level_4_puzzle_index
+    ld a, [w_safari_map_y]
+    ld c, a
+    add a
+    add a
+    add c
+    ld c, a
+    ld a, [w_safari_map_x]
+    add c
+    ld c, a
+    ld a, [w_safari_map_level]
+    add a
+    add a
+    add a
+    add a
+    add c
+    ld c, a
+    ld a, [w_safari_map_level]
+    xor $ff
+    inc a
+    add c
+    add PUZZLE_SAFARI_HAUNTER
```

Comment:
There should be a `FIRST_SAFARI_ZONE_PUZZLE` constant for this, like how pokecrystal has `JOHTO_POKEMON EQU const_value` in pokemon_constants.asm or `KANTO_LANDMARK EQU const_value` in landmark_constants.asm.